src/jit/valuenum.cpp

```diff
@@ -4488,75 +4498,93 @@ void Compiler::fgValueNumberBlock(BasicBlock* blk)
         }
     }
 
-    // Now do the same for "Heap".
-    // Is there a phi for this block?
-    if (blk->bbHeapSsaPhiFunc == nullptr)
-    {
-        fgCurHeapVN = GetHeapPerSsaData(blk->bbHeapSsaNumIn)->m_vnPair.GetLiberal();
-        assert(fgCurHeapVN != ValueNumStore::NoVN);
-    }
-    else
+    // Now do the same for memory.
+    for (MemoryKind memoryKind : allMemoryKinds())
```

Comment:
Do we have to twice the work here, or should we check the bool byrefStatesMatchGcHeapStates